Add async auth guard support for OAuth routes

fahrigedik opened the pull request:
Introduced asyncAuthGuard and asyncAbpOAuthGuard to enable asynchronous authentication checks, particularly for OAuth flows involving code exchange. Updated app routes and OAuth module provider to use the new async guards where appropriate.

### Description

Resolves #23286 (write the related issue number if available)

TODO: Describe what this PR has changed, add screenshot or animated GIF if available, write if it is a breaking change, and how to fix the breaking changes for existing applications if so.

### Checklist

- [x] I fully tested it as developer / designer and created unit / integration tests
- [x] I documented it (or no need to document or I will create a separate documentation issue)

### How to test it?
He describe in #23286 

After setting up the environment for `dev-app`, you can start the application by running the following command in the `npm/ng-packs` directory:
`yarn start`

To reproduce the issue, add the `canActivate `guard to the home route in `npm/ng-packs/apps/dev-app/src/app/app.routes.ts `like this:

```
{
  path: '',
  pathMatch: 'full',
  loadComponent: () => import('./home/home.component').then(m => m.HomeComponent),
  canActivate: [AuthGuard]
}

```

Then, try to access the homepage after logging in — the issue (infinite redirect loop) will occur.

To fix the issue, replace `AuthGuard `with `asyncAuthGuard`.

fahrigedik committed:
Add async auth guard support for OAuth routes

Introduced asyncAuthGuard and asyncAbpOAuthGuard to enable asynchronous authentication checks, particularly for OAuth flows involving code exchange. Updated app routes and OAuth module provider to use the new async guards where appropriate.

sumeyyeKurtulus commented on npm/ng-packs/packages/oauth/src/lib/guards/oauth.guard.ts:
```diff
@@ -53,3 +53,33 @@ export const abpOAuthGuard: CanActivateFn = (
+  state: RouterStateSnapshot,
+) => {
+  const authService = inject(AuthService);
+  const code = new URLSearchParams(window.location.search).get('code');
```
We can use the response type to check whether it is coming from the server since the password flow does not have the same problem:

```tsx
  const { oAuthConfig } = environmentService.getEnvironment();

  if (oAuthConfig?.responseType === 'code') {
    //...
  }
```

sumeyyeKurtulus commented on npm/ng-packs/packages/oauth/src/lib/guards/oauth.guard.ts:
```diff
@@ -53,3 +53,33 @@ export const abpOAuthGuard: CanActivateFn = (
+  if (code) {
+    return firstValueFrom(
+      interval(100).pipe(
+        map(() => authService.isAuthenticated),
```
We should still use `oAuthService.hasValidAccessToken()` here where we can simply use `filter(Boolean)` after that to keep it synced with the other condition and the guard.

sumeyyeKurtulus commented on npm/ng-packs/packages/oauth/src/lib/guards/oauth.guard.ts:
```diff
@@ -53,3 +53,33 @@ export const abpOAuthGuard: CanActivateFn = (
+
+  if (code) {
+    return firstValueFrom(
+      interval(100).pipe(
```
My other suggestion would be using `timer(0, 100)` for immediate and periodic checks. Using  `interval(100)` is obviously not wrong, but it just waits before the first check. For authentication guards, we prefer immediate feedback even if we need a small delay on purpose here.

fahrigedik committed:
Refactor OAuth guard to use OAuthService and timer

Replaces usage of AuthService.isAuthenticated and interval with OAuthService.hasValidAccessToken() and timer for checking authentication status. Also injects EnvironmentService to access OAuth configuration and updates logic to check responseType before waiting for authentication.
